### PR TITLE
Add book page

### PR DESCRIPTION
**Pull request by lubnaabd**
## Description

create addNewBook front-end and back-end code

### Type of change

New features #85 
Description : create book.js that contain fetch function that make a request (post) to back-end and contain validation function.

New features #86 
Description : 
* create book.js in the controller that contain function that deal with the request from the front-end 

New features #16 
Description :
- create getCategory query, that return the all categpries from db.
- create setCategory query, that add all input in db.
- create setBook query, that add all input in db.

### CODE REVIEW

**Review by ali-7 — APPROVED** [review]

File: public/js/book.js [hunk 1]
```diff
@@ -12,14 +26,97 @@ const viewUser = (event, idUser) => {
...
+    const imageBookVal = imageBook.value;
+    const categorySerial = category[category.selectedIndex].value;
+    const data = {
+      nameBookVal,
+      nameAuthorVal,
+      imageBookVal,
+      descriptionVal,
+      categorySerial,
+    };
+    fetch('/admin/books/add', {
```
Comment (RamyAlshurafa) [inline public/js/book.js]: no need to add `add` :)
your method is `POST` that mean you will add a book

File: public/js/book.js [hunk 2]
```diff
@@ -12,14 +26,97 @@ const viewUser = (event, idUser) => {
...
+  const checkNameCategory = check(nameCategory, nameCategoryError, 'هذا الحقل مطلوب');
+  const checkSerialNumber = check(serialNumber, serialNumberCategoryError, 'هذا الحقل مطلوب');
+  if (checkNameCategory && checkSerialNumber) {
+    const nameCategoryVal = nameCategory.value;
+    const serialNumberVal = serialNumber.value;
+    const data = {
+      nameCategoryVal,
+      serialNumberVal,
+    };
+    fetch('/admin/books/add/category', {
```
Comment (RamyAlshurafa) [inline public/js/book.js]: no need for `add` here also

**Review by RamyAlshurafa — CHANGES_REQUESTED** [review]
you made test for `GET` method for `'/admin/books/add'`
same needed for:
- `'/admin/books/add/category'` => `POST`
- `'/admin/books/add'` => `POST`
change routes also

**Review by asmaa-ezz — APPROVED** [review]

File: src/database/queries/set_book.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const dbConnection = require('../db_connection');
+
+const setBook = bookData => new Promise((resolve, reject) => {
+  const {
+    nameBookVal, nameAuthorVal, imageBookVal, descriptionVal, categorySerial,
+  } = bookData;
+  const sql = {
+    text: 'INSERT INTO book(name_book, name_author, image_url, description, category_serial) VALUES($1, $2, $3, $4, $5) RETURNING id,category_serial;    ',
```
Comment (RamyAlshurafa) [inline src/database/queries/set_book.js]: line's length must not exceed 80 characters
do you use linter?

File: src/database/queries/test/test.js [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+  runDbBuild('db_bulid.sql', (err, res) => {
+    t.notOk(err);
+    return runDbBuild('fake_data.sql', () => {
+      const data = { nameCategoryVal: 'تاريخ', serialNumberVal: '507' };
+      setCategory(data)
+        .then((response) => {
+          t.equal(typeof response, 'object', 'setCategory returns data successfully ');
+        })
+        .catch(error => t.error(error));
+      t.end();
```
Comment (RamyAlshurafa) [inline src/database/queries/test/test.js]: move `t.end()` inside `then()`

File: src/database/queries/test/test.js [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+});
+
+test('Test for the setCategory function', (t) => {
+  runDbBuild('db_bulid.sql', (err, res) => {
+    t.notOk(err);
+    return runDbBuild('fake_data.sql', () => {
+      const data = { nameCategoryVal: 'تاريخ', serialNumberVal: '507' };
+      setCategory(data)
+        .then((response) => {
+          t.equal(typeof response, 'object', 'setCategory returns data successfully ');
```
Comment (RamyAlshurafa) [inline src/database/queries/test/test.js]: if you logged your  `response` you will find that it is an empty object

File: src/database/queries/test/test.js [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+    return runDbBuild('fake_data.sql', () => {
+      const data = {
+        nameBookVal: 'يوسف يامريم', nameAuthorVal: 'يامى أحمد', imageBookVal: '', descriptionVal: 'يوسف يامريم', categorySerial: '503',
+      };
+      setBook(data)
+        .then((response) => {
+          t.equal(typeof response, 'object', 'setBook returns data successfully ');
+        })
+        .catch(error => t.error(error));
+      t.end();
```
Comment (RamyAlshurafa) [inline src/database/queries/test/test.js]: move `t.end()` inside `then()`

File: src/database/queries/test/test.js [hunk 7]
```diff
@@ -0,0 +1,53 @@
...
+test('Test for the setBook function', (t) => {
+  runDbBuild('db_bulid.sql', (err, res) => {
+    t.notOk(err);
+    return runDbBuild('fake_data.sql', () => {
+      const data = {
+        nameBookVal: 'يوسف يامريم', nameAuthorVal: 'يامى أحمد', imageBookVal: '', descriptionVal: 'يوسف يامريم', categorySerial: '503',
+      };
+      setBook(data)
+        .then((response) => {
+          t.equal(typeof response, 'object', 'setBook returns data successfully ');
```
Comment (RamyAlshurafa) [inline src/database/queries/test/test.js:36]: if you logged your `response` you will find that it is an empty `object`
so your test always will pass
**be specific to what are you testing**

**Review by RamyAlshurafa — CHANGES_REQUESTED** [review]
## be specific to what are you testing
`t.end()` must be inside `then()`

File: src/database/queries/test/test.js [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+const { setCategory } = require('../set_category');
+const { getCategory } = require('../get_category');
+
+test('Test for the getCategory function', (t) => {
+  runDbBuild('db_bulid.sql', (err, res) => {
+    t.notOk(err);
+    return runDbBuild('fake_data.sql', () => {
+      getCategory()
+        .then((response) => {
+          t.equal(response.length > 0, true, 'getCategory returns data successfully ');
```
Comment (RamyAlshurafa) [inline src/database/queries/test/test.js:14]: you need more tests
you already know the categories in `fake_data` file so you can:
- check the length exactly like  `response.length === 5` .
- check `response[0].name === 'something'`  or  `response[0]["category_serial"] === 'something'`

File: src/database/queries/test/test.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+});
+
+test('Test for the setCategory function', (t) => {
+  runDbBuild('db_bulid.sql', (err, res) => {
+    t.notOk(err);
+    return runDbBuild('fake_data.sql', () => {
+      const data = { nameCategoryVal: 'جغرافتتتjhgjhتيا', serialNumberVal: '511' };
+      setCategory(data)
+        .then((results) => {
+          t.equal(typeof results, 'object', 'setCategory returns data successfully ');
```
Comment (RamyAlshurafa) [inline src/database/queries/test/test.js]: As I said your object here is empty so it's not enough to judge on the function.
you have two ways to be specific:
- Edit `category_serial`'s `sql` and add `RETURNING name` for example, and then check this name.
- Edit the test to be:

```js
test('Test for the setCategory function', (t) => {
  runDbBuild('db_bulid.sql', (err, res) => {
    t.notOk(err);
    return runDbBuild('fake_data.sql', () => {
      const data = { nameCategoryVal: 'جغرافتتتjhgjhتيا', serialNumberVal: '511' };
      getCategory()
        .then((oldResponse) => {
          const oldCategoriesNumber = oldResponse.length;
          setCategory(data)
            .then((results) => {
              getCategory()
                .then((newResponse) => {
                  const newCategoriesNumber = newResponse.length;
                  t.equal(newCategoriesNumber - oldCategoriesNumber, 1, 'setCategory returns data successfully ');
                  t.end();
                });
            })
            .catch(error => t.error(error));
        });
    });
  });
});
```  

Same for `setBook` test

File: src/controllers/books.js [hunk 10]
```diff
@@ -34,14 +38,52 @@ exports.getBorrowedBooks = (request, response) => {
...
+      response.send(JSON.stringify(result));
+    })
+    .catch((err) => {
+      const errorMessage = err.detail;
+      if (err.code === '23505') {
+        const result = { errorMessage: 'This category is Already Exists' };
+        return response.send(JSON.stringify(result));
+      }
+      const result = { errorMessage };
+      return response.send(JSON.stringify(result));
```
Comment (RamyAlshurafa) [inline src/controllers/books.js]: Is this server error? use `response.status(500)`
same with line`70`

**Review by RamyAlshurafa — CHANGES_REQUESTED** [review]
- be more specific with testing
- please don't use the same commit message for various commits

File: src/controllers/books.js (hunk 10 above)
Comment (lubnaabd) [reply src/controllers/books.js]: line 70 I want send error the message from back-end to front-end and alert the message to admin
this category already exists but can replace line 73 response.status(500) or next(err)

Comment (lubnaabd) [reply src/controllers/books.js]: @RamyAlshurafa

Comment (RamyAlshurafa) [reply src/controllers/books.js]: I think if the request sent by fetch you should use `response.status(500)`
but if you use the form to send the request use `next(err)`

**Review by RamyAlshurafa — APPROVED** [review]
Good job :+1: 
For next time: Don't use same commit message for various commits, and  make it more descriptive

**Pull request merged**